code/go/0chain.net/smartcontract/storagesc/challenge.go

```diff
@@ -722,39 +827,77 @@ func (sc *StorageSmartContract) addChallenge(alloc *StorageAllocation,
 
 	var storageChallenge StorageChallenge
 	storageChallenge.ID = challengeID
-	storageChallenge.Validators = selectedValidators
-	storageChallenge.Blobber = selectedBlobberObj
+	//storageChallenge.Validators = selectedValidators
+	storageChallenge.TotalValidators = len(selectedValidators)
+	//storageChallenge.Blobber = blobber
+	storageChallenge.BlobberID = blobberID
 	storageChallenge.RandomNumber = challengeSeed
 	storageChallenge.AllocationID = alloc.ID
-
 	storageChallenge.AllocationRoot = blobberAllocation.AllocationRoot
+	storageChallenge.Created = creationDate
+
+	blobberChallengeObj, err := sc.getBlobberChallenge(blobberID, balances)
+	if err != nil {
+		if err == util.ErrValueNotPresent {
+			blobberChallengeObj = &BlobberChallenge{}
+			blobberChallengeObj.BlobberID = blobberID
+		} else {
+			return "", common.NewError("add_challenge",
+				"error fetching blobber challenge: "+err.Error())
+		}
+	}
 
-	blobberChallengeObj := &BlobberChallenge{}
-	blobberChallengeObj.BlobberID = storageChallenge.Blobber.ID
+	addedChallenge := blobberChallengeObj.addChallenge(&storageChallenge)
+	if !addedChallenge {
+		challengeBytes, err := json.Marshal(storageChallenge)
+		return string(challengeBytes), err
+	}
 
-	blobberChallengeBytes, _ := balances.GetTrieNode(blobberChallengeObj.GetKey(sc.ID))
-	if blobberChallengeBytes != nil {
-		err = blobberChallengeObj.Decode(blobberChallengeBytes.Encode())
-		if err != nil {
-			return "", common.NewError("blobber_challenge_decode_error",
-				"Error decoding the blobber challenge")
+	allocChallengeObj, err := sc.getAllocationChallenge(alloc.ID, balances)
+	if err != nil {
+		if err == util.ErrValueNotPresent {
+			allocChallengeObj = &AllocationChallenge{}
+			allocChallengeObj.AllocationID = alloc.ID
+		} else {
+			return "", common.NewError("add_challenge",
+				"error fetching allocation challenge: "+err.Error())
 		}
 	}
 
-	storageChallenge.Created = creationDate
-	addedChallege := blobberChallengeObj.addChallenge(&storageChallenge)
-	if !addedChallege {
+	addedAllocChallenge := allocChallengeObj.addChallenge(&storageChallenge)
+	if !addedAllocChallenge {
 		challengeBytes, err := json.Marshal(storageChallenge)
 		return string(challengeBytes), err
 	}
 
-	balances.InsertTrieNode(blobberChallengeObj.GetKey(sc.ID), blobberChallengeObj)
+	_, err = balances.InsertTrieNode(allocChallengeObj.GetKey(sc.ID), allocChallengeObj)
```

Comment:
Separate out the following block of code with the  `InsertTrieNode` calls, and put in a separate synchronous second for loop.